### ISSUE TITLE
maps in datadir not accessible with default permissions

### ISSUE THREAD

**Issue by landryb** [georchestra/mapstore2-georchestra#295]
i dunno if its specific to our georchestra configuration, if it has smth to do with #189 - and i remember reporting it in another ticket but i cant find my comment right now.

if i create a new context via the `mapstore/#/admin` webpage, by default it's unaccessible to any users. I need to explicitely grant access to it for all users or a specific group.

unrelated to that, i created a map (not a context) and i wanted to make it available to anyone at `mapstore/#/viewer/openlayers/foo` so i dropped `foo.json` in the georchestra datadir under the mapstore dir (so `/etc/georchestra/mapstore/foo.json` - trying to directly load it (eg access https://my.ids/mapstore/#/viewer/openlayers/foo) i get the unaccessible map error page, and looking at the network console i see that the `/mapstore/rest/config/load/foo.json` gets a 403. Same thing with a non-existent `bar.json` so its as if mapstore didnt collect the maps from the georchestra datadir, even after reloading the webapp.

I know that my map is valid because i can upload foo.json via 'import map' and it works fine.

what is one supposed to do in this case ? since its not a resource stored in the database, i dont think i would need to manually fiddle with the `gs_resource` or `gs_security` database tables..

**Comment by landryb** [comment]
rereading https://mapstore2.readthedocs.io/en/user_docs/developer-guide/maps-configuration/#layers-options im a bit confused by the doc which partly says that those static maps should be edited (pushed?) via the geostore REST API, but static maps could also be taken from the datadir ?

**Comment by landryb** [comment]
eliminated a problem with the sec-proxy by directly talking to the webapp over curl, same 403:
```
curl -v -H 'sec-username: admin' -H 'sec-roles: MAPSTORE_ADMIN' http://localhost:8480/mapstore/rest/config/load/opendata.json
< HTTP/1.1 403
```

getting `new.json` works and returns a code 200, new.json being taken from the webapp root dir as i dont have one in the datadir. Tried putting `foo.json` at the root of the webapp root dir without success either.

**Comment by landryb** [comment]
i like to understand things, so i went looking at the code in https://github.com/geosolutions-it/MapStore2/blob/master/backend/src/main/java/it/geosolutions/mapstore/ConfigController.java#L46 and if i get it right, the backend will return a 403 for all non-whitelisted files (which are only `localConfig, pluginsConfig, extensions` by default) unless one overrides that via `allowed.resources` property - @tdipisa is it what is intended for the georchestra project too ? Or i misunderstood our doc and static maps should be put somewhere else ?

**Comment by landryb** [comment]
bingo. Set the java property in the tomcat env to`-Dallowed.resources=localConfig,pluginsConfig,extensions,config,new,opendata` and my curl request succeeds, and in the end i can open `/#/viewer/openlayers/opendata` in my browser. So by default, no static map can be loaded from the georchestra datadir unless explicitely whitelisted in the `allowed.resources` java property.

**Comment by landryb** [comment]
> i dunno if its specific to our georchestra configuration, if it has smth to do with #189 - and i remember reporting it in another ticket but i cant find my comment right now.
> 
> if i create a new context via the `mapstore/#/admin` webpage, by default it's unaccessible to any users. I need to explicitely grant access to it for all users or a specific group.

ah that was #274

**Comment by catmorales** [comment]
You have to grant right at "everyone" if you want the context be used by all users.

**Comment by catmorales** [comment]
Maps created from  the context manager are stored in the database in the "gs_stored_data" table.

**Comment by landryb** [comment]
yes i know all that but im more about static maps saved in json files and taken from the datadir, not in the database.

re #274 i think having to manually make a map viewable by anyone will be easily forgotten by users creating maps..

**Comment by pierrejego** [comment]
We are also really interested to have context (json file) in datadir. Is there any news .
By the way, is there a doc to upload context (json file) throw an API to the database ?This could be used when context are created in test 
environmentand then publish to prod 
environment

**Comment by catmorales** [comment]
> By the way, is there a doc to upload context (json file) throw an API to the database ?This could be used when context are created in test

Yes, I have the same need to migrate application contexts from the test instance to the production instance.
Copying the records from the test database to the production database is not great.
@tdipisa How do you do it on your side?

**Comment by tdipisa** [comment]
> bingo. Set the java property in the tomcat env to`-Dallowed.resources=localConfig,pluginsConfig,extensions,config,new,opendata` and my curl request succeeds, and in the end i can open `/#/viewer/openlayers/opendata` in my browser. So by default, no static map can be loaded from the georchestra datadir unless explicitely whitelisted in the `allowed.resources` java property.

@landryb I'm sorry for the late reply. Yes, the procedure seems correct. I've asked @offtherailz to check the documentation to include also this information if missing.

@catmorales it is not possible at the moment to upload an application context through the UI of MapStore, in a context wizard at the moment you can only import a map file (step 2 of the wizard). Currently you can use the REST API of MapStore to persist a resource in the DB, therefore also a context (the documentation of available MapStore REST endpoints is available [here](https://github.com/geosolutions-it/geostore/wiki/REST-API) and [here](https://github.com/geosolutions-it/geostore/wiki/REST-API-examples) some samples). If you want we can send you some examples.

**Comment by landryb** [comment]
> @landryb I'm sorry for the late reply. Yes, the procedure seems correct. I've asked @offtherailz to check the documentation to include also this information if missing.

well that feels backwards; from my understanding any json file in the `$datadir/mapstore/` subdir not being `localConfig` or `pluginsConfig` or `extensions` or `config` should be considered a static map.. otherwise that's something else to remember to set at the middleware level, which might not be accessible for some map admins, etc...

**Comment by offtherailz** [comment]
The whitelist is documented here:
https://mapstore.readthedocs.io/en/latest/developer-guide/externalized-configuration/#externalize-front-end-configurations

**Comment by landryb** [comment]
oh right, yes it's documented; but still, so cumbersome. Cant we then have a 'mapstore/maps` subdir with all files whitelisted by default ? mapfishapp had `contexts/` subdir for that and it worked fine...
